src/typechecker.erl

```diff
@@ -3408,14 +3408,54 @@ refinable(_) ->
 no_guards({clause, _, _, Guards, _}) ->
     Guards == [].
 
+
+-spec check_guard_bif(erlang:anno(), atom(), list()) -> map().
+check_guard_bif(P, is_atom, [{var, _, Var}]) -> #{Var => {type, P, atom, []}};
+check_guard_bif(P, is_binary, [{var, _, Var}]) -> #{Var => {type, P, binary, []}};
+check_guard_bif(P, is_bitstring, [{var, _, Var}]) -> #{Var => {type, P, bitstring, []}};
+check_guard_bif(P, is_boolean, [{var, _, Var}]) -> #{Var => {type, P, boolean, []}};
+check_guard_bif(P, is_float, [{var, _, Var}]) -> #{Var => {type, P, float, []}};
+check_guard_bif(P, is_integer, [{var, _, Var}]) -> #{Var => {type, P, integer, []}};
+check_guard_bif(P, is_number, [{var, _, Var}]) -> #{Var => {type, P, number, []}};
+check_guard_bif(P, is_list, [{var, _, Var}]) -> #{Var => {type, P, list, []}};
+check_guard_bif(P, is_map, [{var, _, Var}]) -> #{Var => {type, P, map, []}};
+check_guard_bif(P, is_pid, [{var, _, Var}]) -> #{Var => {type, P, pid, []}};
+check_guard_bif(P, is_port, [{var, _, Var}]) -> #{Var => {type, P, port, []}};
+check_guard_bif(P, is_reference, [{var, _, Var}]) -> #{Var => {type, P, reference, []}};
+check_guard_bif(P, is_tuple, [{var, _, Var}]) -> #{Var => {type, P, tuple, []}};
+check_guard_bif(P, is_function, [{var, _, Var}]) -> #{Var => {type, P, 'fun', []}};
+check_guard_bif(P, is_record, [{var, _, Var}, {atom, _, Record}]) ->
+    #{Var => {type, P, record, [{atom, P, Record}]}};
+check_guard_bif(_P, _Fun, _Vars) ->
+    #{}.
+
+-spec check_guard_call(erlang:anno(), atom(), list()) -> map().
+check_guard_call(P, Name, Args) ->
+    Arity = length(Args),
+    case erl_internal:bif(Name, Arity) of
+        true -> check_guard_bif(P, Name, Args);
+        false -> #{}
```

Comment:
If the call is not to a bif then we should type check it as usual. Call `type_check_expr` here.

Reply:
I think that only guard bifs are allowed in guards. So a false would yield a compiler error. Gradualizer could either emit an error as well, or just ignore the false case (and leave it to other tools).

Reply:
Ah, good point @gomoripeti! I'm fine leaving the code as it is, but it'd be good to add a comment to document the fact that this case should never happen.

Personally, I would very much prefer if arbitrary functions were allowed in guards, except perhaps in receive guards. But that's a debate for another time.

Reply:
I did not know only bifs are allowed in guards. I took a look at all the available ones (Seen here, not sure if up to date https://en.wikibooks.org/wiki/Erlang_Programming/guards#Guard_functions). Might try to expand this to support more of them.

Reply:
This is probably a better list: http://erlang.org/doc/reference_manual/expressions.html#guard-sequences

Reply:
Can we close this? Considering we're only looking for function calls and only specific BIFs are allowed.